pkg/client/client.go

```diff
@@ -205,21 +209,120 @@ func NewClientFromKeys(accessKey, secretAccessKey, region string) (Client, error
 	}, nil
 }
 
+// DescribeRegionsData holds output of DescribeRegions API call and the time when it was last updated.
+type DescribeRegionsData struct {
+	err                   error
+	describeRegionsOutput *ec2.DescribeRegionsOutput
+	lastUpdated           time.Time
+}
+
+type regionCache struct {
+	data  map[string]DescribeRegionsData
+	mutex sync.RWMutex
+}
+
+// RegionCache is accessKeyID (string) to DescribeRegionsData map
+type RegionCache interface {
+	GetCachedDescribeRegions(awsSession *session.Session) (*ec2.DescribeRegionsOutput, error)
+}
+
+// NewRegionCache creates a new empty DescribeRegionsData cache with lock.
+func NewRegionCache() RegionCache {
+	return &regionCache{
+		data:  map[string]DescribeRegionsData{},
+		mutex: sync.RWMutex{},
+	}
+}
+
+// GetCachedDescribeRegions returns DescribeRegionsOutput from DescribeRegions AWS API call.
+// It is cached to avoid AWS API calls on each reconcile loop.
+func (c *regionCache) GetCachedDescribeRegions(awsSession *session.Session) (*ec2.DescribeRegionsOutput, error) {
+	creds, err := awsSession.Config.Credentials.Get()
+	if err != nil {
+		return nil, err
+	}
+
+	c.mutex.Lock()
```

Comment:
We can improve the mutex usage here if you wanted. At the moment you are performing all reads under a write lock, you could obtain the write lock only once you know you need to write.

My suggestion would be to do the read from the map in another helper function that uses RLock, then get the lock in this function only when you determine you need to update it

Reply:
Locking during read here is better because we avoid multiple API requests.

With separate locks, If multiple threads enter at once and there is no data in cache, then they would all wait at write lock and query the API one by one.

With single lock, all threads wait for the first thread to fetch the data and then just read it.

Reply:
There's a pattern for this that we use when using locks, first you read, and decide if you need to write, if you need to write, you wait til you hold the write lock then check again if you need to write, then release the lock if you no longer need to write

But I think we can follow up on this later